File: 백준/Platinum/11376. 열혈강호 2/열혈강호 2.py
from collections import defaultdict

MAX = 1001

tasks = defaultdict(list)
whoWork = [0] * MAX
v = [False] * MAX

def dfs(employee):
    if v[employee]:
        return False
    v[employee] = True
    for n in tasks[employee]:
        if whoWork[n] == 0 or dfs(whoWork[n]):
            whoWork[n] = employee
            return True
    return False

def main():
    import sys
    input = sys.stdin.read
    data = input().splitlines()
    
    N, M = map(int, data[0].split())
    for i in range(1, N + 1):
        line = list(map(int, data[i].split()))
        c = line[0]
        jobs = line[1:]
        tasks[i].extend(jobs)
    
    cnt = 0
    for i in range(1, N + 1):
        for _ in range(2):
            global v
            v = [False] * (N + 1)
            if dfs(i):
                cnt += 1
    
    print(cnt)

if __name__ == "__main__":
    main()

File: 백준/Platinum/11376. 열혈강호 2/열혈강호 2.cc
#include <bits/stdc++.h>
using namespace std;

const int MAX = 1001;

vector<int> tasks[MAX];
int whoWork[MAX];
bool v[MAX];

bool dfs(int employee) {
    if (v[employee])
        return false;
    v[employee] = true;
	for (auto n : tasks[employee]) {
		if (whoWork[n] == 0 || dfs(whoWork[n])) {
			whoWork[n] = employee;
			return true;
		}
	}
	return false;
}

int main() {
	cin.tie(0)->sync_with_stdio(0);
	int N, M;
	cin >> N >> M;
	
	for (int i = 1; i <= N; i++) {
		int c;
		cin >> c;
		for (int j = 0; j < c; j++) {
			int job;
			cin >> job;
			tasks[i].push_back(job);
		}
	}
	int cnt = 0;
	for (int i = 1; i <= N; i++) {
            for (int j=0;j<2;j++) {
                memset(v, 0, N+1);
		        if (dfs(i)) cnt++;
        }
    }

	cout << cnt;
}
